clang/lib/Sema/SemaType.cpp

```diff
@@ -9806,7 +9806,7 @@ QualType Sema::BuildCountAttributedArrayType(QualType WrappedTy,
 /// that expression, according to the rules in C++11
 /// [dcl.type.simple]p4 and C++11 [expr.lambda.prim]p18.
 QualType Sema::getDecltypeForExpr(Expr *E) {
-  if (E->isTypeDependent())
+  if (E->isInstantiationDependent())
```

Comment:
Is it... right? I was assuming this would bring such regressions e.g.

```cpp
template <class T>
const int val = sizeof(T);

template <class T>
void foo() {
  decltype(val<T>);  // <-- this is instantiation-dependent, but not type-dependent.
}
```

Perhaps we can move it to line 9816 where the `PackIndexingExpr` is handled?

Reply:
> Perhaps we can move it to line 9816 where the PackIndexingExpr is handled?

I would be slightly concerned about having specific handling of `PackIndexingExpr` all over the place,
or about  `decltype` behaving differently for pack indexing.

Then again, I'm not sure there is a scenario where it matters.

Reply:
So here is a thought.
If the pack indexing expression is not type dependent, decltype should just produce the type of that expression.
By that logic we do want to modify decltype so that it does something more clever than to try to use the selected expression.

I'll try to do something in that direction, but it's unclear to me under which set of circumstances we can observe that.